File: flight sim EVTOL/scripts/metrics/compute_transition_metrics.py
#!/usr/bin/env python3
"""Transition metrics computation (Python helper for non-MATLAB environments)."""

from __future__ import annotations

from dataclasses import dataclass
import numpy as np


def _first_window(mask: np.ndarray, win: int, start_idx: int = 0):
    for i in range(start_idx, len(mask) - win + 1):
        if np.all(mask[i : i + win]):
            return i
    return None


def compute_transition_metrics(log, target_speed=None, settle_band_frac=0.02, settle_window_s=2.0):
    t = np.asarray(log["t"]).reshape(-1)
    speed = np.asarray(log["speed"]).reshape(-1)
    tilt = np.asarray(log["tilt_deg"]).reshape(-1)
    accel = np.asarray(log.get("accel_xyz", np.zeros((len(t), 3))))
    control = np.asarray(log.get("control", np.zeros((len(t), 1))))

    if target_speed is None:
        target_speed = float(speed[-1])

    dt = np.diff(t)

    # Derive jerk if not provided.
    if "jerk_xyz" in log:
        jerk = np.asarray(log["jerk_xyz"])
    else:
        dacc = np.diff(accel, axis=0)
        jerk = dacc / np.maximum(dt[:, None], 1e-9)
        jerk = np.vstack([jerk, jerk[-1, :]])

    energy_proxy = float(np.trapz(np.sum(np.abs(control), axis=1), t))
    accel_norm = np.linalg.norm(accel, axis=1)
    jerk_norm = np.linalg.norm(jerk, axis=1)
    accel_rms = float(np.sqrt(np.mean(accel_norm**2)))
    jerk_rms = float(np.sqrt(np.mean(jerk_norm**2)))

    if len(t) > 1:
        du = np.diff(control, axis=0)
        urate = du / np.maximum(dt[:, None], 1e-9)
        control_smoothness = float(np.trapz(np.sum(urate**2, axis=1), t[:-1]))
    else:
        control_smoothness = 0.0

    start_threshold = float(np.min(tilt) + 0.05 * (np.max(tilt) - np.min(tilt)))
    starts = np.where(tilt > start_threshold)[0]
    start_idx = int(starts[0]) if len(starts) else 0

    band = settle_band_frac * max(abs(target_speed), 1e-9)
    settled_mask = np.abs(speed - target_speed) <= band

    sample_time = float(np.mean(dt)) if len(dt) else 0.01
    win = max(1, int(round(settle_window_s / max(sample_time, 1e-9))))
    end_idx = _first_window(settled_mask, win, start_idx)

    if end_idx is None:
        transition_time = float("nan")
        settling_time = float("nan")
    else:
        transition_time = float(t[end_idx] - t[start_idx])
        settling_time = transition_time

    overshoot = float(max(0.0, np.max(speed[start_idx:]) - target_speed))

    return {
        "energy_proxy": energy_proxy,
        "accel_rms": accel_rms,
        "jerk_rms": jerk_rms,
        "control_smoothness": control_smoothness,
        "transition_start_s": float(t[start_idx]),
        "transition_time_s": transition_time,
        "overshoot_mps": overshoot,
        "settling_time_s": settling_time,
        "target_speed_mps": float(target_speed),
    }

File: scripts/metrics/compute_transition_metrics.m
function metrics = compute_transition_metrics(logData, opts)
%COMPUTE_TRANSITION_METRICS Compute hover<->cruise transition quality metrics.
%
% Required fields in logData:
%   t           [Nx1] time (s)
%   speed       [Nx1] scalar speed (m/s)
%   tilt_deg    [Nx1] nacelle tilt (deg)
% Optional fields:
%   accel_xyz   [Nx3] acceleration (m/s^2)
%   jerk_xyz    [Nx3] jerk (m/s^3)
%   control     [NxM] actuator command vector
%
% opts (optional):
%   target_speed (default: final speed sample)
%   settle_band_frac (default: 0.02)
%   settle_window_s (default: 2.0)

if nargin < 2
    opts = struct();
end
if ~isfield(opts, 'settle_band_frac'); opts.settle_band_frac = 0.02; end
if ~isfield(opts, 'settle_window_s');  opts.settle_window_s  = 2.0; end

required = {'t','speed','tilt_deg'};
for i = 1:numel(required)
    if ~isfield(logData, required{i})
        error('Missing required field: %s', required{i});
    end
end

t = logData.t(:);
speed = logData.speed(:);
tilt = logData.tilt_deg(:);

if ~isfield(opts, 'target_speed')
    targetSpeed = speed(end);
else
    targetSpeed = opts.target_speed;
end

if isfield(logData, 'accel_xyz')
    accel = logData.accel_xyz;
else
    accel = zeros(numel(t), 3);
end

if isfield(logData, 'jerk_xyz')
    jerk = logData.jerk_xyz;
else
    dt = diff(t);
    dacc = diff(accel, 1, 1);
    jerk = [dacc ./ max(dt, eps), zeros(size(dacc,1), 0)]; %#ok<AGROW>
    jerk = [jerk; jerk(end,:)];
end

if isfield(logData, 'control')
    u = logData.control;
else
    u = zeros(numel(t), 1);
end

% Energy proxy from actuator usage integral.
energyProxy = trapz(t, sum(abs(u), 2));

% Smoothness proxies.
accNorm = sqrt(sum(accel.^2, 2));
jerkNorm = sqrt(sum(jerk.^2, 2));
accRMS = sqrt(mean(accNorm.^2));
jerkRMS = sqrt(mean(jerkNorm.^2));

if size(u,1) > 1
    dt = diff(t);
    du = diff(u, 1, 1);
    uRate = du ./ max(dt, eps);
    ctrlSmooth = trapz(t(1:end-1), sum(uRate.^2, 2));
else
    ctrlSmooth = 0;
end

% Transition start/end heuristics.
startIdx = find(tilt > (min(tilt) + 0.05*(max(tilt)-min(tilt))), 1, 'first');
if isempty(startIdx), startIdx = 1; end

band = opts.settle_band_frac * max(abs(targetSpeed), eps);
settleMask = abs(speed - targetSpeed) <= band;

windowSamps = max(1, round(opts.settle_window_s / max(mean(diff(t)), eps)));
endIdx = NaN;
for k = startIdx:numel(t)-windowSamps+1
    if all(settleMask(k:k+windowSamps-1))
        endIdx = k;
        break;
    end
end

if isnan(endIdx)
    transitionTime = NaN;
    settlingTime = NaN;
else
    transitionTime = t(endIdx) - t(startIdx);
    settlingTime = t(endIdx) - t(startIdx);
end

overshoot = max(0, max(speed(startIdx:end)) - targetSpeed);

metrics = struct();
metrics.energy_proxy = energyProxy;
metrics.accel_rms = accRMS;
metrics.jerk_rms = jerkRMS;
metrics.control_smoothness = ctrlSmooth;
metrics.transition_start_s = t(startIdx);
metrics.transition_time_s = transitionTime;
metrics.overshoot_mps = overshoot;
metrics.settling_time_s = settlingTime;
metrics.target_speed_mps = targetSpeed;

end
